components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java

```diff
@@ -368,6 +376,80 @@ public NotificationResponseBean resendConfirmationCode(User user, Property[] pro
 
     }
 
+    /**
+     * Checks whether the given tenant domain of a username is valid / exists or not.
+     * @param tenantDomain Tenant domain.
+     * @return True if the tenant domain of the user is valid / available, else false.
+     */
+    public boolean isValidTenantDomain(String tenantDomain) {
+
+        boolean isValidTenant = false;
+        try {
+            UserRealm userRealm = getUserRealm(tenantDomain);
+            isValidTenant = userRealm != null;
+
+        } catch (CarbonException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Error while getting user realm for user " + tenantDomain);
+            }
+            // In a case of a non existing tenant.
+            isValidTenant = false;
+        }
+        return isValidTenant;
+    }
+
+    /**
+     * Returns whether a given username is already taken by a user or not.
+     * @param username Username.
+     * @return True if the username is already taken, else false.
+     */
+    public boolean isUsernameAlreadyTaken(String username) {
+
+        boolean isUsernameAlreadyTaken = true;
+        try {
+            String tenantDomain = MultitenantUtils.getTenantDomain(username);
+            String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(username);
+
+            UserRealm userRealm = getUserRealm(tenantDomain);
+            if (userRealm != null) {
+                isUsernameAlreadyTaken = userRealm.getUserStoreManager().isExistingUser(tenantAwareUsername);
+            }
+        } catch (CarbonException | org.wso2.carbon.user.core.UserStoreException e) {
+            if (log.isDebugEnabled()) {
+                log.error("Error while getting user realm for username " + username);
+            }
+            isUsernameAlreadyTaken = true;
```

Comment:
This is not correct. handle this in a different way. Need to show a message to user like "Something went wrong, Please contact system administrator"